### PR TITLE
Fix `ShowWindow(GetConsoleWindow())`

### LINKED ISSUE #13066: [1.14] `ShowWindow(SW_MINIMIZE)` doesn't work anymore
From bug bash. 

1. put in file: 
```pwsh
$Source=@"
using System;
using System.Runtime.InteropServices;
public class Native {
[DllImport("user32.dll")]
public static extern bool ShowWindow(IntPtr hWnd, int nCmdShow);
[DllImport("kernel32.dll", SetLastError = true)]
public static extern IntPtr GetConsoleWindow();
}
"@
Add-Type -TypeDefinition $Source
[Native]::ShowWindow([Native]::GetConsoleWindow(), 6)
```

2. Run `.\hide.ps1`

expect:
* See a dterm sequence for hiding the window

actual:
* Terminal still visible.

### PR DESCRIPTION
**Pull request by zadjii-msft**
A bad merge, that actually revealed a horrible bug.

There was a secret conflict between the code in #12526 and #12515. 69b77ca was a bad merge that hid just how bad the issue was. Fixing the one line `nullptr`->`this` in `InteractivityFactory` resulted in a window that would flash uncontrollably, as it minimized and restored itself in a loop. Great. 

This can seemingly be fixed by making sure that the conpty window is initially created with the owner already set, rather than relying on a `SetParent` call in post. This does pose some complications for the #1256 future we're approaching. However, this is a blocking bug _now_, and we can figure out the tearout/`SetParent` thing in post. 

* fixes #13066.
* Tested with the script in that issue.
* Window doesn't flash uncontrollably.
* `gci | ogv` still works right
* I work here.
* Opening a new tab doesn't spontaneously cause the window to minimize
* Restoring from minimized doesn't yeet focus to an invisible window
* Opening a new tab doesn't yeet focus to an invisible window
* There _is_ a viable way to call `GetAncestor` s.t. it returns the Terminal's hwnd in Terminal, and the console's in Conhost

The `SW_SHOWNOACTIVATE` change is also quite load bearing. With just `SW_NORMAL`, the pseudo window (which is invisible!) gets activated whenever the terminal window is restored from minimized. That's BAD.

There's actually more to this as well. 

Calling `SetParent` on a window that is `WS_VISIBLE` will cause the OS to hide the window, make it a _child_ window, then call `SW_SHOW` on the window to re-show it. `SW_SHOW`, however, will cause the OS to also set that window as the _foreground_ window, which would result in the pty's hwnd stealing the foreground away from the owning terminal window. That's bad.

`SetWindowLongPtr` seems to do the job of changing who the window owner is, without all the other side effects of reparenting the window. 

Without `SetParent`, however, the pty HWND is no longer a descendant of the Terminal HWND, so that means `GA_ROOT` can no longer be used to find the owner's hwnd. For even more insanity, without `WS_POPUP`, none of the values of `GetAncestor` will actually get the terminal HWND. So, now we also need `WS_POPUP` on the pty hwnd. To get at the Terminal hwnd, you'll need

```c++
GetAncestor(GetConsoleWindow(), GA_ROOTOWNER)
```

<details>
<summary>Reparenting</summary>

For #1256.

Although not tested in the scope of the Terminal, I'm fairly confident that `SetWindowLongPtr` will work for this.
image

In that screenshot, I reparented the black terminal to be owned by the blue one. API calls all seemed to confirm that worked, and the hidden pty window wasn't activated as a middle step.

</details>

<details>
<summary>Helper Scripts</summary>

Guys TIL how handy the Win32 interop of powershell is. Look at this. This is so much easier.

```pwsh
$NativeFunctions=@"
using System;
using System.Runtime.InteropServices;
public class Native {

    [DllImport("user32.dll")]
    public static extern bool ShowWindow(IntPtr hWnd, int nCmdShow);

    [DllImport("user32.dll")]
    public static extern IntPtr GetAncestor(IntPtr hWnd, uint gaFlags);

    [DllImport("user32.dll")]
    public static extern IntPtr GetParent(IntPtr hWnd);

    [DllImport("user32.dll")]
    public static extern IntPtr GetWindowLongPtr(IntPtr hWnd, int nIndex);

    [DllImport("user32.dll")]
    public static extern IntPtr SetWindowLongPtr(IntPtr hWnd, int nIndex, IntPtr dwNewLong);

    [DllImport("kernel32.dll", SetLastError = true)]
    public static extern IntPtr GetConsoleWindow();
}
"@
Add-Type -TypeDefinition $NativeFunctions


# [Native]::ShowWindow([Native]::GetConsoleWindow(), 0)

function hwnds {
  $console = [Native]::GetConsoleWindow()

  $GA_PARENT = [Native]::GetAncestor($console, 1) # GA_PARENT
  $GA_ROOT = [Native]::GetAncestor($console, 2) # GA_ROOT
  $GA_ROOTOWNER = [Native]::GetAncestor($console, 3) # GA_ROOTOWNER
  $Parent = [Native]::GetParent($console)
  $GWLP_HWNDPARENT = [Native]::GetWindowLongPtr($console, -8)

  Write-Host ("Console`t`t{0:x}`nGA_PARENT`t{1:x}`nGA_ROOT`t`t{2:x}`nGA_ROOTOWNER`t{3:x}`nParent`t`t{4:x}`nGWLP_HWNDPARENT`t{5:x}" -f $console,$GA_PARENT,$GA_ROOT,$GA_ROOTOWNER,$Parent,$GWLP_HWNDPARENT)

  if ($console -eq $GA_ROOTOWNER) { Write-Host "`e[31mThis is bad`e[m" }
  else {
    if ($console -eq $GA_ROOT) { Write-Host "`e[33mThis is okay`e[m - we told folks that they should use ROOT. They probably will need ROOTOWNER`e[m" }
    else { Write-Host "`e[32mThis is good`e[m" }
  }
}

```

</details>

### CODE REVIEW

File: src/host/VtIo.cpp [hunk 1]
```diff
@@ -300,18 +300,34 @@ bool VtIo::IsUsingVt() const
...
     }
 
     return S_OK;
 }
 
+// Method Description:
+// - Create our pseudowindow. This is exclusively called by
+//   ConsoleInputThreadProcWin32 on the console input thread.
+//    * It needs to be called on that thread, before any other calls to
+//      LocatePseudoWindow, to make sure that the input thread is the HWND's
```
Comment (DHowett) [inline src/host/VtIo.cpp:319]: any way to guarantee that this happens before LPW?

**Comment by DHowett** [comment]
@carlos-zamora FYI last blocker for 1.14!

**Comment by DHowett** [comment]
What horrors happen if a console app uses `SW_NORMAL` on its own window handle inside Terminal? Is there a way for us to suppress activation in the message loop?

**Comment by zadjii-msft** [comment]
> What horrors happen if a console app uses `SW_NORMAL` on its own window handle inside Terminal? Is there a way for us to suppress activation in the message loop?

`[Native]::ShowWindow([Native]::GetConsoleWindow(), 0) ; sleep 1 ; [Native]::ShowWindow([Native]::GetConsoleWindow(), 1)` does seem to work correctly in the terminal, so I suppose this works fine?

**Comment by zadjii-msft** [comment]
🛑 converted to draft for a sec while I make sure that the parenting / ownership stuff didn't regress more than expected.

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 2]
```diff
@@ -1196,8 +1196,11 @@ namespace winrt::Microsoft::Terminal::Control::implementation
 
     void ControlCore::_terminalShowWindowChanged(bool showOrHide)
     {
-        auto showWindow = winrt::make_self<implementation::ShowWindowArgs>(showOrHide);
-        _ShowWindowChangedHandlers(*this, *showWindow);
+        if (_initializedTerminal)
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.cpp:1199]: I would actually bet that code is vestigial at this point. I'd guess leftover from me experimenting with the debouncing. That being said, I thought that 0dc6e0d wasn't needed, and it totally was.

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 3]
```diff
@@ -311,13 +311,18 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
             {
                 THROW_IF_FAILED(ConptyReparentPseudoConsole(_hPC.get(), reinterpret_cast<HWND>(_initialParentHwnd)));
             }
 
+            // GH#12515: The conpty assumes it's hidden at the start. If we're visible, let it know now.
+            if (_initialVisibility)
+            {
+                THROW_IF_FAILED(ConptyShowHidePseudoConsole(_hPC.get(), _initialVisibility));
+            }
+            // THROW_IF_FAILED(ConptyShowHidePseudoConsole(_hPC.get(), _initialVisibility));
```
Comment (DHowett) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp:324]:
```suggestion
```

File: src/cascadia/TerminalConnection/ConptyConnection.h [hunk 4]
```diff
@@ -73,7 +73,7 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
         hstring _commandline{};
         hstring _startingDirectory{};
         hstring _startingTitle{};
-        bool _initialVisibility{ false };
+        bool _initialVisibility{ true };
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalConnection/ConptyConnection.h:76]: i think it makes sense, it was just surprising

**Review by DHowett — APPROVED** [review]
Okay, let's do this. It's dark magic, sure, but so was what we had before. If you worked with @ekoschik on it I have to assume that it could take a bullet and keep going...

**Review by lhecker — APPROVED** [review]

**Pull request merged**